oeps/oep-0018-bp-python-dependencies.rst

```diff
@@ -0,0 +1,344 @@
+  server to perform its primary purpose (``base.in``)
+* Additional dependencies which are only needed when optional extra features
+  of a package are desired
+* Assorted testing libraries to run automated test suites (``test.in``)
+* Static code analysis tools to perform code quality checks (``quality.in``)
+* The utilities called directly by a CI server to create and use one or more
+  virtualenvs and report code coverage statistics to a 3rd-party service
+  (``jenkins.in``, ``travis.in``)
+* `Sphinx`_ and other utilities used to generate developer documentation
+  (``doc.in``)
+* Additional utilities needed to perform common development tasks (``dev.in``)
+* Utilities that a particular developer likes to use with a repository, but
+  aren't strictly needed for any of the regular contexts (``private.in``).
+
+.. _Sphinx: http://www.sphinx-doc.org/
+
+Declare Direct Dependencies
+---------------------------
+
+As indicated above, some of the usage contexts have a standard filename used in
+the ``requirements`` directory of an Open edX repository to list dependencies.
+Others will have an appropriate filename custom to that repository's unique
+context.  Each of these is a ``pip``-compatible `requirements file`_ listing
+the direct dependencies needed for that context.  Beyond complying with the
+file format, there are a few guidelines each of these files should follow:
+
+* The file should start with a brief comment explaining the context in which
+  these dependencies are needed.  Examples can be found in the
+  `cookiecutter-django-app`_ repository.
+* Each listed dependency should have a brief end-of-line comment explaining
+  its primary purpose(s) in this context.  These comments typically start at
+  the 27th character, but this is just a convention for consistency with files
+  generated by ``pip-compile``.
+* Version constraints should only be used to exclude dependency versions which
+  are known (or strongly suspected) to not work in this context.
+* Indirect dependencies (used by dependencies but not directly by the code in
+  the repository itself) should not be listed unless a constraint is needed to
+  enforce a compatible version; these are automatically detected and captured
+  elsewhere as described below.
+* `Environment markers`_ should be used as necessary to indicate dependencies
+  which should only be installed on specific operating systems, Python
+  versions, etc.
+* If the dependencies in one context are a superset of those in another one,
+  do not repeat the dependencies.  Instead, explicitly include the file
+  listing the common dependencies in the superset context file (e.g.
+  ``-r base.txt``) and explain in an end-of-line comment why that set of
+  dependencies is also needed in this context.  Note that the ``pip-compile``
+  output file should be included rather than the looser requirements file it
+  was generated from, as this ensures that the same versions of packages are
```

Comment:
I'm confused by the "include the output file rather than the requirements file."  An example would help.

Reply:
Yeah, this is getting kind of long and hard to follow for a bullet point.  I'll see if I can clarify it a bit.